### PR TITLE
Add order note before do_action() to maintain order of notes

### PR DESCRIPTION
**Pull request by menakas**
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

While doing a status transition of an order in the [status_transition() function](https://github.com/woocommerce/woocommerce/blob/43f6e67bdf3de4590a57a465eaa1b1baf26458fe/includes/class-wc-order.php#L355), an order note is added [here](https://github.com/woocommerce/woocommerce/blob/43f6e67bdf3de4590a57a465eaa1b1baf26458fe/includes/class-wc-order.php#L377) after the `woocommerce_order_status_changed` action. When we hook into this action, and do another status transition(as in Subscriptions), this causes a lot of confusion because the order note for the second status transition is added before the first one.

This order note should be added before the `do_action()` calls [here](https://github.com/woocommerce/woocommerce/blob/43f6e67bdf3de4590a57a465eaa1b1baf26458fe/includes/class-wc-order.php#L369-L370) so that any note added by functions that hook into these actions do not get added before the status transition order note. 

### How to test the changes in this Pull Request:

1. Add a function that hooks into `woocommerce_order_status_changed`. Let this function add an order note. Something like this:
```
        add_action( 'woocommerce_order_status_changed', 'my_func_adds_order_note', 10, 4 );
        function my_func_adds_order_note( $order_id, $from_status, $to_status, $order ) {
	     $order->add_order_note( 'Can I change status to something else?', 'woocommerce');
	     return;
         }
```

2. Create an order and change its status.
3. On master it can be noted that the order of notes is such that the status change note appears after the note in the `my_func_adds_order_note` like [this](https://cld.wthms.co/LEzqCB). On this branch, the status change note appears first and then the note in the function.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Add order note immediately after status change before the 'woocommerce_order_status_changed' action.

### CODE REVIEW

**Comment by hahn208** [comment]
I like this commit because it would fix my issue with trying to add the order note to the order status email.

`$order = wc_get_order( $order_id );`
`$order->update_status('completed', "Tracking number: ABC123");`

In this scenario I do not have access to the order note for the completed email with the provided hooks.

`do_action( 'woocommerce_order_status_' . $status_transition['from'] . '_to_' . $status_transition['to'], $this->get_id(), $this );`

`do_action( 'woocommerce_order_status_changed', $this->get_id(), $status_transition['from'], $status_transition['to'], $this );`

$status_transition['note'] is not provided for some random reason and the note hasn't been saved yet.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Not a fan of having this code duplicated, probably better to create a private method for add the note.
Suggestion for the method:

```php
private function add_status_transition_note( $note, $transition ) {
	return $this->add_order_note( trim( $transition['note'] . ' ' . $note ), 0, $transition['manual'] );
}
```

**Comment by menakas** [comment]
Yes. Added the function and pushed a commit.

File: includes/class-wc-order.php [hunk 1]
```diff
@@ -1705,6 +1708,18 @@ public function add_order_note( $note, $is_customer_note = 0, $added_by_user = f
 		return $comment_id;
 	}
 
+	/**
+	 * Add an order note for status transition
+	 *
```
Comment (claudiosanches) [inline includes/class-wc-order.php]: Please, just include `@since 3.9.0` and everything will be good to merge.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
Just one more tweak and will be good to merge.
Thanks for your help!

File: includes/class-wc-order.php (hunk 1 above)
Comment (menakas) [reply includes/class-wc-order.php]: Done, thanks.

**Review by claudiosanches — APPROVED** [review]
Thanks for your contribution!

**Pull request merged**